# Introduction

This project works with a dataset of used cars from *eBay Kleinanzeigen*, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The dataset was originally [scraped](https://en.wikipedia.org/wiki/Web_scraping) and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). But a few modifications was made by the dataquest team from the original dataset that was uploaded to Kaggle:

The dataset was dirtied a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

The data dictionary provided with data is as follows:

- **dateCrawled** - When this ad was first crawled. All field-values are taken from this date.
- **name** - Name of the car.
- **seller** - Whether the seller is private or a dealer.
- **offerType** - The type of listing
- **price** - The price on the ad to sell the car.
- **abtest** - Whether the listing is included in an A/B test.
- **vehicleType** - The vehicle Type.
- **yearOfRegistration** - The year in which the car was first registered.
- **gearbox** - The transmission type.
- **powerPS** - The power of the car in PS.
- **model** - The car model name.
- **odometer** - How many kilometers the car has driven.
- **monthOfRegistration** - The month in which the car was first registered.
- **fuelType** - What type of fuel the car uses.
- **brand** - The brand of the car.
- **notRepairedDamage** - If the car has a damage which is not yet repaired.
- **dateCreated** - The date on which the eBay listing was created.
- **nrOfPictures** - The number of pictures in the ad.
- **postalCode** - The postal code for the location of the vehicle.
- **lastSeenOnline** - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

Let's start by importing the libraries we need and reading the dataset into pandas.

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Reading the data into a pandas dataframe
auto = pd.read_csv('../Dataset/autos.csv', encoding = 'Windows-1252')

In [3]:
auto

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
auto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

The above output shows that there is at maximum, 50000 rows in our dataset. But some columns show that there are `null` values within them. Such columns as `vehicleType`, `gearbox`, `model`, `fuelType` and `notRepairedDamage` have less than 50k non-null values.

There are five numeric columns in the dataset and 15 string columns, amounting the total columns in the dataset to 20.

None of the columns have more than ~20% null values.

The column names use [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case), which means we can't just replace spaces with underscores.

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [5]:
auto.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We'll copy this output so we can make a few edits to some columns names and use snake case for all column names

In [6]:
cleaned_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen']

auto.columns = cleaned_columns
auto.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Above we cleaned the columns of our dataset manually because the columns used camel case, so we couldn't use functions to replace strings and insert underscores betweeen words.

In [7]:
auto.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-14 20:50:02  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  no_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

The above output shows the descriptives of all columns in the dataset. At inspection, the column `no_of_pictures` column doesn't look very good and will lead us to dropping it because it may be of no use for any further analysis.

The `price`, `odometer` columns are of the string dtype when they should be numeric type. Thus, these columns will need further cleaning.

The `offer_type` and `seller` columns indicates that there are two unique values in them but the frequencies shows just one value occuring 49,999 times, thus there is no need for these columns as it will be of no use to our analysis 

In [8]:
auto['no_of_pictures'].value_counts()

0    50000
Name: no_of_pictures, dtype: int64

Further investigation of the `no_of_pictures` column shows that only one value `0` appears 50k times,

In [9]:
auto['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

Further investigation of the seller column shows that `private` seller occurs 49,999 times while `gewerblich` appeared just once.

In [10]:
auto['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

The above shows that `Angebot` also occupies almost all the rows in the `offer_type column`. Thus, we'll drop these columns from our dataset.

In [11]:
auto = auto.drop(columns = ['no_of_pictures', 'seller', 'offer_type'])

In [12]:
auto['price'].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

By inspecting the `price` column further we noticed that all values follow a particular pattern, which is the attachment of the dollar sign **$**, so we'll need to remove this sign from items in the column

In [13]:
auto['price'] = (auto['price']
                 .str
                 .replace('$', '')
                 .str
                 .replace(',', '')
                 .astype(float))

auto = auto.rename(columns = {'price':'price_usd'})


In [14]:
auto['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

On inspection also we found that the `odometer` column is supposed to hold numeric values but the string `km` invalidates that. We'll remove the string and convert the values to floats

In [15]:
auto['odometer'] = (auto['odometer']
                   .str
                   .replace('km', '')
                   .str
                   .replace(',', '')
                   .astype(float))

In [16]:
auto['odometer']

0        150000.0
1        150000.0
2         70000.0
3         70000.0
4        150000.0
           ...   
49995    100000.0
49996    150000.0
49997      5000.0
49998     40000.0
49999    150000.0
Name: odometer, Length: 50000, dtype: float64

In [17]:
auto = auto.rename(columns = {'odometer':'odometer_km'}) #remane the odometer column

We'll look further into the `odometer_km` and `price_usd` columns to find any outliers and remove them from the column values

In [18]:
auto['price_usd'].unique().shape

(2357,)

The above output shows that there are 2,357 unique prices for cars in our dataset.

In [19]:
auto['price_usd'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

The e notation shows that the values are large and follows the logic of `5*10 ^ 4` for the count output and `9.840044*10^3` for the mean output and so on... There are 50k rows in the `price_usd` column and the average price of cars in the dataset is $9,840.044.

We observe from the output that cars with the the highest prices cost 100,000,000 dollars and cars with the least prices are free (0.00). We also observe that 75 percent of cars in our dataset cost below 7,200 dollars, while 50 percent of cars in the dataset cost below 2,950 dollars.

In [20]:
auto['price_usd'].value_counts().sort_index(ascending = False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price_usd, dtype: int64

Above we see the top five most expensive cars and find that a particular car costs 99,999,999 dollars. One slightly odd pattern in the prices is above is that, they increase steadily at reasonable intervals upto 350k dollars, before the prices jumped unrealistically to 999,990k dollars.



In [21]:
auto['price_usd'].value_counts().sort_index(ascending = True).head(20)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price_usd, dtype: int64


There are 1,421 free listed cars - we might consider removing these rows. Also about 30 listings costs less than 30 dollars and 14 listings costs above 1million dollars. We will remove cars costing above 350k dollars and cars that cost 0 dollars

In [22]:
auto = auto[(auto['price_usd'] != 0.0) 
           &
           (auto['price_usd'] <= 350000.0)]
auto['price_usd'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_usd, dtype: float64

In [23]:
auto['odometer_km'].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

The average distance covered by cars in the dataset is 125,770.1 kilometers. The minimum distance covered by a listed car is 5k kilometers while the maximum distance is 150k kilometers. Also the 75th percentile shows that 50% of listed cars have driven less than 150k kilometers. Since 150k kilometers is the maximum distance covered by cars in the dataset, this invariably means that 50% of the cars have driven at least 150k kilometers.

Therefore, majority of the cars in the dataset have traveled a distance of at least 150k kilometers.

##### Exploring date columns in the dataframe

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself.

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

lets explore these columns further to see the kind of data they hold

In [24]:
auto['date_crawled'].head()

0    2016-03-26 17:47:46
1    2016-04-04 13:38:56
2    2016-03-26 18:57:24
3    2016-03-12 16:58:10
4    2016-04-01 14:38:50
Name: date_crawled, dtype: object

In [25]:
auto['last_seen'].head()

0    2016-04-06 06:45:54
1    2016-04-06 14:45:08
2    2016-04-06 20:15:37
3    2016-03-15 03:16:28
4    2016-04-01 14:38:50
Name: last_seen, dtype: object

In [26]:
auto['ad_created'].head()

0    2016-03-26 00:00:00
1    2016-04-04 00:00:00
2    2016-03-26 00:00:00
3    2016-03-12 00:00:00
4    2016-04-01 00:00:00
Name: ad_created, dtype: object

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. Also we notice that the first 10 characters represent the day **(e.g. 2016-03-12)**. To understand the date range, we can extract just the date values

In [27]:
auto['registration_month'].head()

0    3
1    6
2    7
3    6
4    7
Name: registration_month, dtype: int64

In [28]:
auto['registration_year'].head()

0    2004
1    1997
2    2009
3    2007
4    2003
Name: registration_year, dtype: int64

The other two columns `registration_month` and `registration_year` are represented as numeric values, so we can use methods like `Series.describe()` to understand the distribution without any extra data processing.

In [29]:
(auto['date_crawled']
.str[:10]
.value_counts(normalize = True, dropna = False)
.sort_index()
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period beginning from March 5th 2016 to April 7th 2016. If we sort the values in descending order to get the dates with the highest frequency relative to other dates, we'll get the following output:

In [30]:
(auto['date_crawled']
.str[:10]
.value_counts(normalize = True, dropna = False)
.sort_values(ascending = False)
)

2016-04-03    0.038608
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-12    0.036920
2016-03-14    0.036549
2016-04-04    0.036487
2016-03-07    0.036014
2016-04-02    0.035478
2016-03-28    0.034860
2016-03-19    0.034778
2016-03-15    0.034284
2016-03-29    0.034099
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-22    0.032987
2016-03-11    0.032575
2016-03-23    0.032225
2016-03-26    0.032204
2016-03-10    0.032184
2016-03-31    0.031834
2016-03-17    0.031628
2016-03-25    0.031607
2016-03-27    0.031092
2016-03-16    0.029610
2016-03-24    0.029342
2016-03-05    0.025327
2016-03-13    0.015670
2016-03-06    0.014043
2016-04-05    0.013096
2016-03-18    0.012911
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

Above we see that the distribution of listings crawled on each day is roughly uniform, with `2016-04-03` having the highest frequency of cars crawled

In [31]:
(auto['last_seen']
.str[:10]
.value_counts(normalize = True, dropna = False)
.sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The `last_seen` column most likely depicts when last the crawler saw this ad online. We could assume that this date shows when the car was sold and therefore had its ad removed.

The last three dates shows that the proportion of cars seen online is 6 - 10x more than the proportion of other dates. This is likely because the crawling period is ending

In [32]:
(auto['ad_created']
.str[:10]
.value_counts(normalize = True, dropna = False)
.sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

The above output shows when the ad was created. The difference between the first date and last date is about 10 months. Most fall within months when the listings was crawled, that is, between march and april and a smaller proportion was created at more backward dates.

In [33]:
auto['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

In [34]:
auto['registration_year'].value_counts().head(60)

2000    3156
2005    2936
1999    2897
2004    2703
2003    2699
2006    2670
2001    2636
2002    2486
1998    2363
2007    2277
2008    2215
2009    2085
1997    1951
2011    1623
2010    1589
2017    1392
1996    1373
2012    1310
1995    1227
2016    1220
2013     803
2014     663
1994     629
2018     470
1993     425
2015     392
1992     370
1990     347
1991     339
1989     174
1988     135
1985      95
1980      85
1987      72
1986      72
1984      51
1983      51
1978      44
1982      41
1970      38
1979      34
1972      33
1981      28
1968      26
1967      26
1971      26
1974      24
1973      23
1960      23
1977      22
1966      22
1976      21
1969      19
1975      18
1965      17
1964      12
1963       8
1959       6
1961       6
1910       5
Name: registration_year, dtype: int64

Above we see that values from the year 1910 show fewer cars. We may decide to remove years less than 1900 since they will contain very few number of cars.

Lets get the percentage of cars with dates less than 1900, i.e between the least date 1000 and 1900 while excluding 1900

In [35]:
((~auto['registration_year'].between(1900, 2016)).sum() / auto.shape[0] * 100)

3.8793369710697

The '`tilde` **~**' sign helps us to return opposite values of what we really computed. The above output is suppose to show the percentage of values between the year 1900 and 2016, but with the **tilde** operator **(~)** we get the opposite of what we actually computed, that is the percentage of values that fall outside this range (less than 1900 and more than 2016).

Alternatively, we can find the values separately and still arrive at same answer by doing the following

In [36]:
# get year values below the expected year
a = (auto['registration_year'].between(1000, 1900, inclusive = False)).sum() / auto.shape[0]

# get year values above the expected year
b = (auto['registration_year'].between(2016, 9999, inclusive = False)).sum() / auto.shape[0]
(a+b) * 100

3.871100586842376

From the above output, we can conclude that values outside the range of 1900 to 2016 holds less than 4% of the data in our dataset. Thus removing them won't have any severe effects on the dataset.

In [37]:
# subset the dataset to get year values within the bounds of the dataset
auto = auto[auto['registration_year'].between(1900, 2016)]

In [38]:
auto['registration_year'].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

The above output shows that there are now 46,681 rows in our dataset. The minimum registration year is 1910 while the maximum registration year is 2016. The 25th percentile shows that 75% of the cars were registered after after the year 2000.

In [39]:
auto['brand'].describe()

count          46681
unique            40
top       volkswagen
freq            9862
Name: brand, dtype: object

We found that the volkswagen brand has the most cars in the dataset.

Lets explore further to find other car brands that represent a significant proportion of cars in our dataset.

In [40]:
(auto['brand'].value_counts(normalize = True)) * 100

volkswagen        21.126368
bmw               11.004477
opel              10.758124
mercedes_benz      9.646323
audi               8.656627
ford               6.989996
renault            4.714980
peugeot            2.984083
fiat               2.564212
seat               1.827296
skoda              1.640925
nissan             1.527388
mazda              1.518819
smart              1.415994
citroen            1.400998
toyota             1.270324
hyundai            1.002549
sonstige_autos     0.981127
volvo              0.914719
mini               0.876159
mitsubishi         0.822604
honda              0.784045
kia                0.706926
alfa_romeo         0.664082
porsche            0.612669
suzuki             0.593389
chevrolet          0.569825
chrysler           0.351321
dacia              0.263490
daihatsu           0.250637
jeep               0.227073
subaru             0.214220
land_rover         0.209936
saab               0.164949
jaguar             0.156381
daewoo             0

Above we found that volkswagen cars represent ~21% of the entire cars, bmw has ~11%, opel ~11%, mercedes_benz ~10%, audi ~9%, ford ~7% while renault has ~ 5% of cars in the dataset. We'll be working with these seven brands because each hold at least 5% of cars in the dataset.

In [41]:
# get the counts of brands
count_brands = auto['brand'].value_counts(normalize = True) * 100

# get the major brands
major_brands = count_brands[count_brands >= 4.7].index
print(major_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford',
       'renault'],
      dtype='object')


Above we selected the brands that holds approximately 5% of cars in our dataset. and assigned them to the variable `major_brands`

In [42]:
common_brands = {}

for brand in major_brands:
    selected_brands = auto[auto['brand'] == brand]
    mean_price = selected_brands['price_usd'].mean()
    common_brands[brand] = round(mean_price, 2)
    
import operator

list_of_brands = sorted(common_brands.items(), key = operator.itemgetter(1), reverse = True)

highest_priced_brands = dict(list_of_brands)

highest_priced_brands

{'audi': 9336.69,
 'mercedes_benz': 8628.45,
 'bmw': 8332.82,
 'volkswagen': 5402.41,
 'ford': 3749.47,
 'opel': 2975.24,
 'renault': 2474.86}

The above output shows the average prices of the most common car brands in the dataset. Audi cars are priced highest on the average, followed by mercedes_benz cars. On the average, Renault cars are priced least amongst the top 7 car brands in the dataset. 

Volkswagen cars are in-between highest priced brands and least priced brands amongst the top 7 most popular car brands.

In [43]:
# Create a dictionary that holds the average distance covered for the top 7 car brands
mileage = {}

for brand in major_brands:
    selected_brands = auto[auto['brand'] == brand]
    mean_mileage = selected_brands['odometer_km'].mean()
    mileage[brand] = round(mean_mileage, 2)
    
most_mileage = sorted(mileage.items(), key = operator.itemgetter(1), reverse = True)

highest_mileage_brands = dict(most_mileage)

highest_mileage_brands

{'bmw': 132572.51,
 'mercedes_benz': 130788.36,
 'opel': 129310.04,
 'audi': 129157.39,
 'volkswagen': 128707.16,
 'renault': 128071.33,
 'ford': 124266.01}

In [44]:
# Create a series object from the two dictionaries
prices = pd.Series(highest_priced_brands)
miles_covered = pd.Series(highest_mileage_brands)
print(prices)
miles_covered

audi             9336.69
mercedes_benz    8628.45
bmw              8332.82
volkswagen       5402.41
ford             3749.47
opel             2975.24
renault          2474.86
dtype: float64


bmw              132572.51
mercedes_benz    130788.36
opel             129310.04
audi             129157.39
volkswagen       128707.16
renault          128071.33
ford             124266.01
dtype: float64

In [45]:
# create a dataframe of the top 7 car brands and their average prices
price_and_mileage = pd.DataFrame(prices, columns = ['mean_prices'])
price_and_mileage

mean_prices
audi               9336.69
mercedes_benz      8628.45
bmw                8332.82
volkswagen         5402.41
ford               3749.47
opel               2975.24
renault            2474.86

In [46]:
# add a new column that holds the average miles covered to the newly created dataframe
price_and_mileage['miles_covered'] = miles_covered

In [47]:
price_and_mileage

mean_prices  miles_covered
audi               9336.69      129157.39
mercedes_benz      8628.45      130788.36
bmw                8332.82      132572.51
volkswagen         5402.41      128707.16
ford               3749.47      124266.01
opel               2975.24      129310.04
renault            2474.86      128071.33

We can deduce from the above output that cars with the highest prices don't necessarily have the most mileage. The car brand with the least price in the top 7 has more mileage than 5th most priced brand.